File: src/core/data_formats/asm/data_flow_analysis/get_map_designations.ts
import { InstructionSegment } from "../instructions";
import { ControlFlowGraph } from "./ControlFlowGraph";
import { OP_BB_MAP_DESIGNATE, OP_MAP_DESIGNATE, OP_MAP_DESIGNATE_EX } from "../opcodes";
import { get_register_value } from "./get_register_value";
import { LogManager } from "../../../Logger";

const logger = LogManager.get("core/data_formats/asm/data_flow_analysis/map_designations");

export function get_map_designations(
    instruction_segments: InstructionSegment[],
    func_0_segment: InstructionSegment,
): Map<number, number> {
    const map_designations = new Map<number, number>();
    let cfg: ControlFlowGraph | undefined;

    for (const inst of func_0_segment.instructions) {
        switch (inst.opcode.code) {
            case OP_MAP_DESIGNATE.code:
            case OP_MAP_DESIGNATE_EX.code:
                {
                    if (!cfg) {
                        cfg = ControlFlowGraph.create(instruction_segments);
                    }

                    const area_id = get_register_value(cfg, inst, inst.args[0].value);

                    if (area_id.size() !== 1) {
                        logger.warning(`Couldn't determine area ID for map_designate instruction.`);
                        continue;
                    }

                    const variant_id_register =
                        inst.args[0].value + (inst.opcode.code === OP_MAP_DESIGNATE.code ? 2 : 3);
                    const variant_id = get_register_value(cfg, inst, variant_id_register);

                    if (variant_id.size() !== 1) {
                        logger.warning(
                            `Couldn't determine area variant ID for map_designate instruction.`,
                        );
                        continue;
                    }

                    map_designations.set(area_id.get(0)!, variant_id.get(0)!);
                }
                break;

            case OP_BB_MAP_DESIGNATE.code:
                map_designations.set(inst.args[0].value, inst.args[2].value);
                break;
        }
    }

    return map_designations;
}

File: lib/src/commonMain/kotlin/world/phantasmal/lib/assembly/dataFlowAnalysis/GetMapDesignations.kt
package world.phantasmal.lib.assembly.dataFlowAnalysis

import mu.KotlinLogging
import world.phantasmal.lib.assembly.InstructionSegment
import world.phantasmal.lib.assembly.OP_BB_MAP_DESIGNATE
import world.phantasmal.lib.assembly.OP_MAP_DESIGNATE
import world.phantasmal.lib.assembly.OP_MAP_DESIGNATE_EX

private val logger = KotlinLogging.logger {}

fun getMapDesignations(
    instructionSegments: List<InstructionSegment>,
    func0Segment: InstructionSegment,
): Map<Int, Int> {
    val mapDesignations = mutableMapOf<Int, Int>()
    var cfg: ControlFlowGraph? = null

    for (inst in func0Segment.instructions) {
        when (inst.opcode.code) {
            OP_MAP_DESIGNATE.code,
            OP_MAP_DESIGNATE_EX.code,
            -> {
                if (cfg == null) {
                    cfg = ControlFlowGraph.create(instructionSegments)
                }

                val areaId = getRegisterValue(cfg, inst, inst.args[0].value as Int)

                if (areaId.size > 1) {
                    logger.warn {
                        "Couldn't determine area ID for ${inst.opcode.mnemonic} instruction."
                    }
                    continue
                }

                val variantIdRegister =
                    inst.args[0].value as Int + (if (inst.opcode == OP_MAP_DESIGNATE) 2 else 3)
                val variantId = getRegisterValue(cfg, inst, variantIdRegister)

                if (variantId.size > 1) {
                    logger.warn {
                        "Couldn't determine area variant ID for ${inst.opcode.mnemonic} instruction."
                    }
                    continue
                }

                mapDesignations[areaId[0]!!] = variantId[0]!!
            }

            OP_BB_MAP_DESIGNATE.code -> {
                mapDesignations[inst.args[0].value as Int] = inst.args[2].value as Int
            }
        }
    }

    return mapDesignations
}
